src/VoiceProtocol.cpp

```diff
@@ -0,0 +1,57 @@
+// Copyright 2005-2020 The Mumble Developers. All rights reserved.
+// Use of this source code is governed by a BSD-style license
+// that can be found in the LICENSE file at the root of the
+// Mumble source tree or at <https://www.mumble.info/LICENSE>.
+
+#include "VoiceProtocol.h"
+
+#include <string>
+
+VoiceProtocol::VoiceProtocol(const std::string &s)
+	: m_valid(false), m_protocolName(s), m_transport(VoiceTransportType::INVALID) {
+}
+
+VoiceProtocol::VoiceProtocol() : m_valid(false), m_transport(VoiceTransportType::INVALID) {
+}
+
+std::string VoiceProtocol::toString() const {
+	return m_protocolName;
+}
+
+std::shared_ptr< VoiceProtocol > VoiceProtocol::fromString(const std::string &s) {
+	if (s.rfind("MUMBLE_UDP_", 0) != std::string::npos)
```

Comment:
Why `rfind`? If this is an official protocol, we expect the string to start with this prefix. In that case regular `find` should do the trick. And shouldn't we checking whether the string starts with this prefix instead of checking whether it contains it? :thinking:

Reply:
Yeah, `find` will do the job. `rfind` finds the string before or at the specified position, in this case, 0. I made it here just to make sure we are matching the prefix. The other possibility is, as you have mentioned, to use `find` and check if the location is at 0. I can also do this if it is more intuitive.

Reply:
Yeah I think checking ` find() == 0` is easier to grasp